src/Core/Style.js

```diff
@@ -79,44 +79,51 @@ function readVectorProperty(property, options) {
-    const target = document.createElement('img');
-
-    if (typeof source == 'string') {
-        if (value) {
-            const color = new Color(value);
-            Fetcher.texture(source, { crossOrigin: 'anonymous' })
-                .then((texture) => {
-                    const img = texture.image;
-                    canvas.width = img.naturalWidth;
-                    canvas.height = img.naturalHeight;
-                    const ctx = canvas.getContext('2d', { willReadFrequently: true });
-                    ctx.drawImage(img, 0, 0);
-                    const imgd = ctx.getImageData(0, 0, img.naturalWidth, img.naturalHeight);
-                    const pix = imgd.data;
-
-                    const colorToChange = new Color('white');
-                    for (let i = 0, n = pix.length; i < n; i += 4) {
-                        const d = deltaE(pix.slice(i, i + 3), colorToChange) / 100;
-                        pix[i] = (pix[i] * d +  color.r * 255 * (1 - d));
-                        pix[i + 1] = (pix[i + 1] * d +  color.g * 255 * (1 - d));
-                        pix[i + 2] = (pix[i + 2] * d +  color.b * 255 * (1 - d));
-                    }
-                    ctx.putImageData(imgd, 0, 0);
-                    target.src = canvas.toDataURL('image/png');
-                });
+async function loadImage(source) {
+    let promise = cacheStyle.get(source, 'null');
+    if (!promise) {
+        promise = Fetcher.texture(source, { crossOrigin: 'anonymous' });
+        cacheStyle.set(promise, source, 'null');
+    }
+    return (await promise).image;
+}
+
+function drawImage(img, cropValues = { width: img.naturalWidth, height: img.naturalHeight }) {
+    canvas.width = cropValues.width;
+    canvas.height = cropValues.height;
+    const ctx = canvas.getContext('2d', { willReadFrequently: true });
+    ctx.drawImage(img,
+        cropValues.x || 0, cropValues.y || 0, cropValues.width, cropValues.height,
+        0, 0, cropValues.width, cropValues.height);
+    return ctx.getImageData(0, 0, cropValues.width, cropValues.height);
+}
+
+async function getImage(source, options) {
+    if (!options.cropValues && !options.color) {
+        return source;
+    }
+    const img = options.img || await loadImage(source);
+    let imgd = drawImage(img, options.cropValues);
+    if (options.color) {
+        const imgdColored = cacheStyle.get(options.id || source, options.color);
+        if (!imgdColored) {
+            const pix = imgd.data;
+            const color = new Color(options.color);
+            const colorToChange = new Color('white');
+            for (let i = 0, n = pix.length; i < n; i += 4) {
+                const d = deltaE(pix.slice(i, i + 3), colorToChange) / 100;
+                pix[i] = (pix[i] * d +  color.r * 255 * (1 - d));
+                pix[i + 1] = (pix[i + 1] * d +  color.g * 255 * (1 - d));
+                pix[i + 2] = (pix[i + 2] * d +  color.b * 255 * (1 - d));
+            }
```

Comment:
Would it be possible to encapsulate my coloring method?

Reply:
I moved it all in a function. Is it better ?